### PR TITLE
ROCm-5.0: version bumps for compiler stack

### PR DESCRIPTION
**Pull request by littlewu2508**
This PR bumps llvm-roc, rocm-device-libs, rocm-comgr, rocminfo, hip to 5.0.2. 

@candrews  @justxi

There is quite a lot of changes, most importantly is I move the install prefix of hip from `/usr/lib/hip` to `/usr`, because we currently don't allow multi slot of hip, and install directly into `/usr` simplify the packaging of downstream projects (such as tensorflow). Many downstream packages assume hipcc is installed into `${ROCM_PATH}/bin` while we make `ROCM_PATH=/usr`.

https://github.com/ROCm-Developer-Tools/hipamd/issues/18 is another bug I encounter, and I give a solution without much sane.

### CODE REVIEW

**Comment by heroxbd** [comment]
We are missing `dev-libs/rocr-runtime-5`.

**Comment by heroxbd** [comment]
You could check this branch with `pkgcheck scan --commits`.

**Comment by heroxbd** [comment]
`5.0.2-update-header.patch: No such file or directory` from the HIP ebuild.

**Comment by littlewu2508** [comment]
> 

It's a large patch, should be a distfile

**Comment by littlewu2508** [comment]
> `5.0.2-update-header.patch: No such file or directory` from the HIP ebuild.

It's a large patch, should be a distfile

**Comment by heroxbd** [comment]
> > `5.0.2-update-header.patch: No such file or directory` from the HIP ebuild.
> 
> It's a large patch, should be a distfile

But it is not in the distfile.

**Comment by heroxbd** [comment]
> > > `5.0.2-update-header.patch: No such file or directory` from the HIP ebuild.
> > 
> > 
> > It's a large patch, should be a distfile
> 
> But it is not in the distfile.

It fails if you do not enable USE=profile.

**Comment by heroxbd** [comment]
If I run `pkgcheck scan --commits`, I see

```
dev-libs/rocr-runtime  NonexistentDeps: version 9999: DEPEND: nonexistent package: >=sys-devel/llvm-roc-9999                   
dev-libs/rocr-runtime  NonexistentDeps: version 9999: RDEPEND: nonexistent package: >=sys-devel/llvm-roc-9999

hip:  UnusedInherits: version 5.0.2: unused eclasses: flag-o-matic, multilib
llvm-roc: VariableScope: version 5.0.2: variable 'EROOT' used in 'src_install', line 92
```

Please fix them.  Drop `dev-libs/rocr-runtime` in a different commit.

Squash the related hip commits.

**Comment by heroxbd** [comment]
> Upstream move to hipamd.git, while the perl scripts is in original HIP.git, wieh .pl suffix. Future hipcc will be a binary.

typo: Upstream moves to hipamd.git, while the perl scripts is in original HIP.git with .pl suffix.  Future hipcc will be a binary.

**Comment by heroxbd** [comment]
Is `/usr/lib64/libhsakmt.so.1.0.6` not used by `hip` anymore?  I see hip still links to it.  Why remove it from DEPEND?

File: dev-libs/rocm-comgr/rocm-comgr-5.0.2.ebuild [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+LICENSE="MIT"
+SLOT="0/$(ver_cut 1-2)"
+
+RDEPEND=">=dev-libs/rocm-device-libs-${PV}
+	>=sys-devel/llvm-roc-${PV}:="
+DEPEND="${RDEPEND}"
+
+CMAKE_BUILD_TYPE=Release
+
+# PATCHES=( "${FILESDIR}"/${PN}-5.0-rocm_path.patch )
```
Comment (heroxbd) [inline dev-libs/rocm-comgr/rocm-comgr-5.0.2.ebuild:33]: Remove it.

File: dev-libs/rocm-comgr/rocm-comgr-5.0.2.ebuild [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+src_prepare() {
+	sed '/sys::path::append(HIPPath/s,"hip","",' -i src/comgr-env.cpp || die
+	sed '/sys::path::append(LLVMPath/s,"llvm","lib/llvm/roc",' -i src/comgr-env.cpp || die
+	sed '/Args.push_back(HIPIncludePath/,+1d' -i src/comgr-compiler.cpp || die
+	sed '/Args.push_back(ROCMIncludePath/,+1d' -i src/comgr-compiler.cpp || die # ROCM and HIPIncludePath is now /usr, which disturb the include order
+	eapply "${FILESDIR}"/${PN}-5.0-rocm_path.patch
+	eprefixify src/comgr-env.cpp
```
Comment (heroxbd) [inline dev-libs/rocm-comgr/rocm-comgr-5.0.2.ebuild:41]: try `eapply $(prefixify_ro "${FILESDIR}"/${PN}-5.0-rocm_path.patch)`

File: sys-devel/llvm-roc/llvm-roc-5.0.2.ebuild [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		-DLLVM_ENABLE_DOXYGEN=OFF
+		-DLLVM_INSTALL_UTILS=ON
+		-DLLVM_VERSION_SUFFIX=roc
+		-DOCAMLFIND=NO
+	)
+
+	use debug || local -x CPPFLAGS="${CPPFLAGS} -DNDEBUG"
+
+	cmake_src_configure
+}
```
Comment (heroxbd) [inline sys-devel/llvm-roc/llvm-roc-5.0.2.ebuild:87]: Running a diff against llvm-4.5.2 resulted
```
 src_configure() {                                                                                        
-       PROJECTS="clang;lld;llvm"                                                                         
+       PROJECTS="clang;lld"                         
                                                                                                          
        if usex runtime; then                        
                PROJECTS+=";compiler-rt"             
        fi                                                                                                
                                                                                                          
        local mycmakeargs=(                                                                               
-               -Wno-dev                                                                                  
                -DCMAKE_INSTALL_PREFIX="${EPREFIX}/usr/lib/llvm/roc"                                      
                -DCMAKE_DISABLE_FIND_PACKAGE_CUDA=ON                                                      
                -DLLVM_ENABLE_PROJECTS="${PROJECTS}"                                                      
                -DLLVM_TARGETS_TO_BUILD="AMDGPU;X86"                                                      
                -DLLVM_BUILD_DOCS=NO                                                                      
-               -DLLVM_ENABLE_BINDINGS=OFF                                                                
                -DLLVM_ENABLE_OCAMLDOC=OFF           
                -DLLVM_ENABLE_SPHINX=NO              
                -DLLVM_ENABLE_DOXYGEN=OFF                                                                 
-               -DLLVM_BUILD_UTILS=ON                                                                     
+               -DLLVM_INSTALL_UTILS=ON                                                                   
                -DLLVM_VERSION_SUFFIX=roc                                                           
                -DOCAMLFIND=NO                       
        )                                 
```
Could you please explain them in the commit message?

**Review by heroxbd — CHANGES_REQUESTED** [review]

File: sys-devel/llvm-roc/llvm-roc-5.0.2.ebuild [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+PATCHES=(
+	"${FILESDIR}/${PN}-4.2.0-current_pos.patch"
+	"${FILESDIR}/${PN}-5.0.0-linkdl.patch"
+)
+
+CMAKE_BUILD_TYPE=RelWithDebInfo
+
+src_prepare() {
+	cd "${WORKDIR}/llvm-project-rocm-${PV}" || die
+	# eapply "${FILESDIR}/${PN}-3.0.0-add_libraries.patch"
```
Comment (heroxbd) [inline sys-devel/llvm-roc/llvm-roc-5.0.2.ebuild:34]: Remove it.

**Comment by heroxbd** [comment]
Please explain why we should bundle `rocclr` and `roctracer` with HIP.

File: dev-libs/rocm-comgr/rocm-comgr-5.0.2.ebuild (hunk 2 above)
Comment (littlewu2508) [reply dev-libs/rocm-comgr/rocm-comgr-5.0.2.ebuild:41]: > Is `/usr/lib64/libhsakmt.so.1.0.6` not used by `hip` anymore? I see hip still links to it. Why remove it from DEPEND?

Emm, hip's CMakeLists.txt doesn't have any related information about hsakmt; only roctracer requires libhsakmt.so. With `USE=-profile`, there is no sign of `hsakmt` in `build.ninja`. If force removed roct-thunk-interface, the compile will not fail (but hip program fails at runtime). I guess the compiler is automatically link libhsakmt.so even if libhsakmt.so is not present.

**Comment by heroxbd** [comment]
littlewu2508 ***@***.***> writes:

> but hip program fails at runtime

This judges it to go into `RDEPEND` .

**Comment by littlewu2508** [comment]
On Sun, Mar 20, 2022 at 09:16:51AM -0700, Benda Xu wrote:
> Could you please explain them in the commit message?
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/gentoo/gentoo/pull/24681#pullrequestreview-915122647
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

Actually I start developing llvm-roc-5 based on 4.3.0-r1. It seems that these changes is introduced when llvm-roc is bumped to 4.5.2 in https://github.com/gentoo/gentoo/pull/23343. My observation is, these changes make no big difference.

--

File: sys-devel/llvm-roc/llvm-roc-5.0.2.ebuild (hunk 3 above)
Comment (littlewu2508) [reply sys-devel/llvm-roc/llvm-roc-5.0.2.ebuild:87]: > Running a diff against llvm-4.5.2 resulted
> 
> ```
>  src_configure() {                                                                                        
> -       PROJECTS="clang;lld;llvm"                                                                         
> +       PROJECTS="clang;lld"                         
>                                                                                                           
>         if usex runtime; then                        
>                 PROJECTS+=";compiler-rt"             
>         fi                                                                                                
>                                                                                                           
>         local mycmakeargs=(                                                                               
> -               -Wno-dev                                                                                  
>                 -DCMAKE_INSTALL_PREFIX="${EPREFIX}/usr/lib/llvm/roc"                                      
>                 -DCMAKE_DISABLE_FIND_PACKAGE_CUDA=ON                                                      
>                 -DLLVM_ENABLE_PROJECTS="${PROJECTS}"                                                      
>                 -DLLVM_TARGETS_TO_BUILD="AMDGPU;X86"                                                      
>                 -DLLVM_BUILD_DOCS=NO                                                                      
> -               -DLLVM_ENABLE_BINDINGS=OFF                                                                
>                 -DLLVM_ENABLE_OCAMLDOC=OFF           
>                 -DLLVM_ENABLE_SPHINX=NO              
>                 -DLLVM_ENABLE_DOXYGEN=OFF                                                                 
> -               -DLLVM_BUILD_UTILS=ON                                                                     
> +               -DLLVM_INSTALL_UTILS=ON                                                                   
>                 -DLLVM_VERSION_SUFFIX=roc                                                           
>                 -DOCAMLFIND=NO                       
>         )                                 
> ```

@inode64 Could you explain those changes when bumping llvm-roc from 4.3.0-r1 to 4.5.2? I think both of the configurations are OK.

**Comment by littlewu2508** [comment]
On Sun, Mar 20, 2022 at 09:20:32AM -0700, Benda Xu wrote:
> Please explain why we should bundle `rocclr` and `roctracer` with HIP.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/gentoo/gentoo/pull/24681#issuecomment-1073284527
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>
Done. INSTALL.md[https://github.com/ROCm-Developer-Tools/hipamd/blob/develop/INSTALL.md] indicate that they should be build togeher.
--

**Pull request closed without merging**